destination_middleware.go

```diff
@@ -236,3 +253,143 @@ func (d *destinationWithRateLimit) Write(ctx context.Context, recs []Record) (in
 	}
 	return d.Destination.Write(ctx, recs)
 }
+
+// -- DestinationWithRecordFormat ----------------------------------------------
+
+const (
+	configDestinationRecordFormat        = "sdk.record.format"
+	configDestinationRecordFormatOptions = "sdk.record.format.options"
+)
+
+// DestinationWithRecordFormat adds support for changing the output format of
+// records, specifically of the Record.Bytes method. It adds two parameters to
+// the destination config:
+//   - `sdk.record.format` - The format of the output record. The inclusion
+//     validation exposes a list of valid options.
+//   - `sdk.record.format.options` - Options are used to configure the format.
+type DestinationWithRecordFormat struct {
+	// DefaultRecordFormat is the default record format.
+	DefaultRecordFormat string
+	RecordFormatters    []RecordFormatter
```

Comment:
I agree, it would be easier for us internally, but it would be harder for the implementer to define record formatters. This struct is supposed to be used by the destination connector implementer if they want to change record formatters (e.g. remove a formatter they don't want, add a custom formatter to the list). In that case it's easier to do this:
```
sdk.DestinationWithRecordFormat {
	RecordFormatters: []sdk.RecordFormatter{
		sdk.TemplateFormatter{},
		MyCustomFormatter{},
	}
}
```
Than this:
```
sdk.DestinationWithRecordFormat {
	RecordFormatters: map[string]sdk.RecordFormatter{
		sdk.TemplateFormatter{}.Name(): sdk.TemplateFormatter{},
		MyCustomFormatter{}.Name(): MyCustomFormatter{},
	}
}
```

We could internally change this slice into a map and store it in a private field, but I thought why bother if we can solve it with a binary search, it's not a performance critical function and we don't expect tons of formatters.

Reply:
Ah, right, it will be used "from outside" as well. An `Add` method may provide a balance between a simpler internal implementation and ease of use, but initializing the field looks good too. :+1: